### PR TITLE
Dismiss gesture hint when gesture is abandoned on scroll

### PR DESCRIPTION
**Pull request by onurpolattimur**
fixes: #1242

### CODE REVIEW

**Review by raineorshine — CHANGES_REQUESTED** [review]
Thanks! It is an improvement, though it still happens sometimes unfortunately. It's even harder to reproduce now. I can reproduce it most consistently when the scrollbar is completely still at the start, and the final down swipe is just a few pixels long. If you are unable to reproduce then I will have to get remote debugging working and try to track the touchmove events.

Architecturally, we need to take a slightly different approach. The `MultiGesture` component should not know about the global store or gesture hints. I suggest adding a new event handler `onCancel` that allows the upstream component to add the appropriate logic.

**Comment by onurpolattimur** [comment]
I could reproduce only once but I made some changes. Can you check it again, please?

File: src/components/MultiGesture.tsx [hunk 1]
```diff
@@ -87,6 +88,9 @@ class MultiGesture extends React.Component<MultiGestureProps> {
 
       onPanResponderMove: (e: GestureResponderEvent, gestureState: GestureState) => {
         if (this.abandon) {
+          if (this.props.onCancel) {
+            this.props.onCancel()
+          }
```
Comment (raineorshine) [inline src/components/MultiGesture.tsx:93]: You can also use the fancy new [Optional Chaining](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Operators/Optional_chaining) operator: `this.props.onCancel?.()`

**Review by raineorshine — COMMENTED** [review]
It still occurs consistently for me unfortunately. Would you like to keep working on it or shall I give it a try?

**Comment by onurpolattimur** [comment]
> It still occurs consistently for me unfortunately. Would you like to keep working on it or shall I give it a try?

I can't re-produce still after 8384ab9. So you can give it a try if you want.

**Comment by raineorshine** [comment]
@onurpolattimur The main issue was the scrolling was causing the `MultiGesture` to terminate without `onPanResponderRelease` (and thus `onEnd`) being called.

In order to make the gesture system more forgiving, we currently allow a tiny bit of scroll without abandoning the gesture. Unfortunately, in the case of this issue, `onPanResponderRelease` is not called, even after `touchend`. Neither is `onPanResponderReject` or `onPanResponderEnd`.

`onPanResponderTerminate` is called consistently, but it is also called for any scroll event. I am not aware of a way to differentiate when `onPanResponderTerminate` is called from a scroll event vs a final termination where release it never called.

So instead of eliminating the scroll lenience, we can handle `touchend` manually and ensure `onEnd` is called if there is a valid sequence.

```ts
    document.body.addEventListener('touchend', e => {
      if (this.sequence) {
        setTimeout(() => {
          if (!this.abandon && this.sequence) {
            this.props.onEnd?.(this.sequence, e as unknown as GestureResponderEvent)
            this.reset()
          }
        })
      }
    })
```

I also cleaned up some state access that was happening outside of a thunk, which could lead to timing issues, but I don't think that was causing the problem.

**Comment by raineorshine** [comment]
Apparently I can't request a review from you since you created the PR, but feel free to take a look.

**Comment by onurpolattimur** [comment]
Hi @raineorshine. Glad you found it.
I reviewed and it looks good to me.

**Review by raineorshine — APPROVED** [review]

**Pull request merged**